feat: add accessor arrays support to `blas/ext/base/gcusumpw`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add accessor arrays support to `blas/ext/base/gcusumpw`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add accessor arrays support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/index.d.ts:
```diff
@@ -20,7 +20,17 @@
+type InputArray = NumericArray | Collection<number> | AccessorArrayLike<number>;
+
+/**
+* Input array.
```
```suggestion
* Output array.
```

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/index.d.ts:
```diff
@@ -44,7 +54,7 @@ interface Routine {
 	* // y => [ 1.0, -1.0, 1.0 ]
 	*/
-	( N: number, sum: number, x: NumericArray, strideX: number, y: NumericArray, strideY: number ): NumericArray;
+	( N: number, sum: number, x: InputArray, strideX: number, y: OutputArray, strideY: number ): OutputArray;
```
```suggestion
	<T extends OutputArray>( N: number, sum: number, x: InputArray, strideX: number, y: T, strideY: number ): T;
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/index.d.ts:
```diff
@@ -66,7 +76,7 @@ interface Routine {
 	* // y => [ 1.0, -1.0, 1.0 ]
 	*/
-	ndarray( N: number, sum: number, x: NumericArray, strideX: number, offsetX: number, y: NumericArray, strideY: number, offsetY: number ): NumericArray;
+	ndarray( N: number, sum: number, x: InputArray, strideX: number, offsetX: number, y: OutputArray, strideY: number, offsetY: number ): OutputArray;
```
```suggestion
	ndarray<T extends OutputArray>( N: number, sum: number, x: InputArray, strideX: number, offsetX: number, y: T, strideY: number, offsetY: number ): T;
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/test.ts:
```diff
@@ -26,7 +27,8 @@ import gcusumpw = require( './index' );
 
-	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType NumericArray
+	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType OutputArray
+	gcusumpw( x.length, 0.0, new AccessorArray( x ), 1, new AccessorArray( y ), 1 ); // $ExpectType OutputArray
```
```suggestion
	gcusumpw( x.length, 0.0, new AccessorArray( x ), 1, new AccessorArray( y ), 1 ); // $ExpectType AccessorArray
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/test.ts:
```diff
@@ -139,7 +141,8 @@ import gcusumpw = require( './index' );
 
-	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType NumericArray
+	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType OutputArray
+	gcusumpw.ndarray( x.length, 0.0, new AccessorArray( x ), 1, 0, new AccessorArray( y ), 1, 0 ); // $ExpectType OutputArray
```
```suggestion
	gcusumpw.ndarray( x.length, 0.0, new AccessorArray( x ), 1, 0, new AccessorArray( y ), 1, 0 ); // $ExpectType AccessorArray
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/test.ts:
```diff
@@ -26,7 +27,8 @@ import gcusumpw = require( './index' );
 	const y = new Float64Array( 10 );
 
-	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType NumericArray
+	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType OutputArray
```
```suggestion
	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType Float64Array
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/test.ts:
```diff
@@ -139,7 +141,8 @@ import gcusumpw = require( './index' );
 	const y = new Float64Array( 10 );
 
-	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType NumericArray
+	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType OutputArray
```
```suggestion
	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType Float64Array
```

headlessNode committed:
refactor: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented:
@kgryte changes applied.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/test.ts:
```diff
@@ -26,7 +27,8 @@ import gcusumpw = require( './index' );
 
-	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType NumericArray
+	gcusumpw( x.length, 0.0, x, 1, y, 1 ); // $ExpectType Float64Array
+	gcusumpw( x.length, 0.0, new AccessorArray( x ), 1, new AccessorArray( y ), 1 ); // $ExpectType AccessorArray
```
```suggestion
	gcusumpw( x.length, 0.0, new AccessorArray( x ), 1, new AccessorArray( y ), 1 ); // $ExpectType AccessorArray<number>
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/docs/types/test.ts:
```diff
@@ -139,7 +141,8 @@ import gcusumpw = require( './index' );
 
-	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType NumericArray
+	gcusumpw.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType Float64Array
+	gcusumpw.ndarray( x.length, 0.0, new AccessorArray( x ), 1, 0, new AccessorArray( y ), 1, 0 ); // $ExpectType AccessorArray
```
```suggestion
	gcusumpw.ndarray( x.length, 0.0, new AccessorArray( x ), 1, 0, new AccessorArray( y ), 1, 0 ); // $ExpectType AccessorArray<number>
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/lib/accessors.js:
```diff
@@ -0,0 +1,112 @@
+* @returns {Object} output array object
+*
+* @example
+* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
+* var arraylike2object = require( '@stdlib/array/base/arraylike2object' );
+*
+* var x = toAccessorArray( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
+* var y = toAccessorArray( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+*
+* gcusumpw( 4, 0.0, arraylike2object( x ), 2, 1, arrayLike2object( y ), 1, 0 );
```
While this example is correct, it doesn't fulfill the secondary role of unit testing. Ideally, we'd illustrate expected behavior.

```suggestion
* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
* var arraylike2object = require( '@stdlib/array/base/arraylike2object' );
*
* var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ];
* var y = [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ];
*
* gcusumpw( 4, 0.0, arraylike2object( toAccessorArray( x ) ), 2, 1, arrayLike2object( toAccessorArray( y ) ), 1, 0 );
* // y => [ 2.0, 3.0, 5.0, 3.0, 1.0, 3.0, 6.0, 10.0 ]
```

This likely also applies to other similar PRs.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/lib/ndarray.js:
```diff
@@ -62,13 +64,20 @@ var BLOCKSIZE = 128;
+	ox = arraylike2object( x );
+	oy = arraylike2object( y );
+	if ( ox.accessorProtocol || oy.accessorProtocol ) {
+		return accessors( N, sum, ox, strideX, offsetX, oy, strideY, offsetY );
```
Incorrect return value.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumpw/test/test.main.js:
```diff
@@ -128,6 +128,94 @@ tape( 'the function computes the cumulative sum', function test( t ) {
+
 tape( 'the function returns a reference to the output array', function test( t ) {
 	var out;
 	var x;
```
Accessor array test case here and in the `ndarray` tests.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte reviewed (approved):
LGTM
